windows_registry.py: Improve registry search efficiency, error reporting

johnwparent opened the pull request:
Windows registry searches for keys matching a regex were previously recursive by default. Introduce ability to limit depth to 1 level of keys.

Dramatically improve error handling/reporting and clarify scenarios where certain errors are expected/acceptable. This is mostly useful for the frequently esoteric `OSError` that can be raised during Windows Registry interactions. When encountering unexpected errors, report key path, subkeys, values, and registry query parameters.

Adds debug output for all registry interactions

Improves handling of race conditions where a key is opened, but another process removes the key before it can be read.

johnwparent committed:
Windows registry: helpful interactions

johnwparent committed:
reg test stub

johnwparent committed:
Update reg logic

johnwparent committed:
catch missed refactor

johnwparent committed:
Refactor winreg interface

johnwparent committed:
Improve error handling

johnwparent committed:
Reduce search by several orders of magnitude

johnwparent committed:
Remove extraneous use of winreg

johnwparent committed:
Harden compiler detection against registry failures

johnwparent committed:
Remove testing stub

johnwparent commented:
@spackbot fix style

johnwparent committed:
[@spackbot] updating style on behalf of johnwparent

scheibelp commented on lib/spack/spack/operating_systems/windows_os.py:
```diff
@@ -83,11 +85,19 @@ def compiler_search_paths(self):
+        except OSError as e:
+            tty.debug(
+                f'Windows registry query on "SOFTWARE\\WOW6432Node\\Microsoft" under HKEY_LOCAL_MACHINE: {str(e)}'
+            )
```
A couple questions:

* Should this retry on failure?
* Is there a way to distinguish race-condition-type errors from other errors (e.g. perhaps permissions errors that ought to percolate up to the user)?

johnwparent replied:
> Should this retry on failure?

I don't see why not, I'll implement 3 retries.

> Is there a way to distinguish race-condition-type errors from other errors

Unfortunately not AFAICT, from my testing, it seems like `winreg` returns the same error regardless  of the context. I think the "error" that could potentially percolate up to the user is just them failing to find a compiler they know is in the registry and correlating with this debug message.

scheibelp commented on lib/spack/spack/util/windows_registry.py:
```diff
@@ -280,25 +346,11 @@ def _traverse_subkeys(self, stop_condition, collect_all_matching=False):
-        return self._traverse_subkeys(search_key, collect_all_matching=collect_all_matching)
-
-    def find_subkey(self, subkey_name):
+    def find_subkey(self, subkey_name, depth=True):
```
Update docstring with new arg. If I understand correctly, this determines whether you list the first level or whether you recurse into subkeys. Typically we have used "recursive" for this notion, which I think is also clearer in this context.

johnwparent replied:
Sounds good, amended.

scheibelp commented on lib/spack/spack/util/windows_registry.py:
```diff
@@ -59,18 +59,88 @@ def subkeys(self):
+                raise e
+            # Other OS errors are more difficult to diagnose, so we wrap them in some extra
+            # reporting
+            raise InvalidRegistryOperation("OpenKeyEx", e, self.path, subname, **kwargs) from e
```
I see some repetition between this and `EnumKey`/`EnumValue`: would it be worth collecting this logic into a common object with a `__call__` method? This might be a naive observation, and e.g. if the reused error codes are circumstantial then I could see the value in keeping them separate (possibly with an extra comment though).

johnwparent replied:
It's generalizable, I think another method in the class implementing this logic is probably sufficient.

scheibelp commented on lib/spack/spack/util/windows_registry.py:
```diff
@@ -59,18 +59,88 @@ def subkeys(self):
+            return winreg.OpenKeyEx(self.hkey, subname, **kwargs)
+        except OSError as e:
+            # Expected errors that occur on occasion, these are easily
+            # debug-able and have sufficiently verbose reporting and obvious cause
```
IMO it's worth briefly explaining what these error codes mean (which in my case would help me understand why they don't apply for `QueryInfoKey`, for example).

johnwparent replied:
I've elaborated on the specific easily diagnoseable errors (and consolidated this similar logic as suggested above).

It's not that these don't apply to `QueryInfoKey`, it's that the errors raised by `QueryInfoKey` have only been observed to raise the vague, difficult to discern errors the others do only on occasion, and it interacts with the keys differently (reading metadata vs interacting with they keys directly). The other errors (`[WinError [2|5]`) are fairly obvious w.r.t. how/why they occurred (file does not exist, and permission errors) but while thinking about it, while I'm centralizing the logic, I might as well include this and future proof.

scheibelp reviewed (changes requested):
Needs style update. In the meantime I have some refactoring suggestions. Let me know if these seem off track.

johnwparent committed:
windows os: multiple directory tries

johnwparent committed:
style update

johnwparent committed:
centralize error logic

johnwparent committed:
the almighty style check

scheibelp commented on lib/spack/spack/operating_systems/windows_os.py:
```diff
@@ -83,11 +85,25 @@ def compiler_search_paths(self):
+                        'Windows registry query on "SOFTWARE\\WOW6432Node\\Microsoft"'
+                        f"under HKEY_LOCAL_MACHINE: {str(e)}"
+                    )
+                tries += 1
```
A couple questions:

* Is this likely to succeed if re-executed immediately? Should this sleep for a short while?
* How do programs generally handle this periodic unavailability?

johnwparent replied:
I will need to investigate both of these questions. Will get back to you as soon as I have reasonable answers (in my initial testing, only one system (out of three) succeeded on an immediate retry) and am able to look into how other programs call out to the registry.

johnwparent replied:
So on further investigation, the slate of unexpected errors (those outside of `FileNotFound` and `InvalidPermissions`) can be raised for a couple of reasons, some of which are race conditions, but others of which are (AFAIU, I read a couple forum threads) effectively corner cases of permission issues where some programs can write Keys that are "readable" for your user given the ACL for the entry, but are actually locked.

Given that in some cases this is never going to go away (we just can't read the key), then I think we absolutely don't want to sleep here. As I understand it, the race conditions are generally caused by some monitor software (anti-malware or otherwise) reading the keys and should be resolved more or less instantaneously. 

All of the python reg interactions I was able to find on GH implement more or less the same approach we do here, `try/except` blocks.

With all that in mind, I think we should not sleep, and only retry once. Thoughts?

scheibelp commented on lib/spack/spack/operating_systems/windows_os.py:
```diff
@@ -83,11 +85,25 @@ def compiler_search_paths(self):
+                vs_entries = msft.find_subkeys(r"VisualStudio_.*", depth=False)
+                break
+            except OSError as e:
+                if tries == 2:
```
For readability, can you replace `tries -> failures` and do this check after the increment so it looks like:

```
failures += 1
# IMO this reads more-naturally: "we failed 3 times"
if failures == 3:
  ...
```

johnwparent replied:
Makes sense to me!

scheibelp commented on lib/spack/spack/operating_systems/windows_os.py:
```diff
@@ -83,11 +85,25 @@ def compiler_search_paths(self):
+                )
+                vs_entries = msft.find_subkeys(r"VisualStudio_.*", depth=False)
+                break
+            except OSError as e:
```
I think it would be worth adding a comment about the inability to distinguish race conditions from other types of errors like permission issues.

johnwparent replied:
Good call! I'm going to look a little deeper into the win32 reg api the python library abstracts and see if there's maybe a mechanism that can improve the error reporting as well.

johnwparent replied:
Will update comment with results of my research as mentioned in: https://github.com/spack/spack/pull/41720#discussion_r1475152144

scheibelp commented on lib/spack/spack/util/windows_registry.py:
```diff
@@ -59,49 +59,111 @@ def subkeys(self):
+            # Expected errors that occur on occasion, these are easily
+            # debug-able and have sufficiently verbose reporting and obvious cause
+            if hasattr(e, "winerror") and e.winerror in (5, 2):
+                raise e
```
A couple things:

* this appears to duplicate logic from `_handle_winreg_errors`
* some repetition could be avoided by defining a wrapper/context-manager for

```
        try:
            ...
        except OSError as e:
            self._handle_winreg_errors(e, "EnumKey", index)
```

like

```
with self.error_handler():
  return winreg.QueryValueEx(self.hkey, name, **kwargs)
```

johnwparent replied:
Great suggestion, will implement.

scheibelp reviewed (changes requested):
Most comments here are minor.

I was curious about the likely success rate of immediate retries in `compiler_search_paths`.

johnwparent committed:
WinReg: refactor error handling to use context manager

johnwparent committed:
WindowsOs: Update reg query re-try logic

johnwparent committed:
Update comments explaining winreg behavior

johnwparent commented:
@scheibelp I believe I have addressed all comments. In particular the changes to the req query in WindowsOs could use another look (I implemented what I thought best based on my comment but I didn't wait for your response so I realize I'll probably need to refactor)

johnwparent committed:
Style

johnwparent committed:
WindowsOS: remove over-indent

scheibelp commented on lib/spack/spack/util/windows_registry.py:
```diff
@@ -280,55 +327,45 @@ def _traverse_subkeys(self, stop_condition, collect_all_matching=False):
         Return:
             the desired subkey as a RegistryKey object, or none
 
         For more details, see the WindowsRegistryView._find_subkey_s method docstring
```
The `_find_subkey_s` method was removed (I assumed because it just directly calls another function and that's it? seems like a good reason to me but just want to make sure I understand why).

johnwparent replied:
IIRC yes that's why, and we folded the "more info" into these method docstrings anyway, so it was facile for a number of reasons.

scheibelp commented on lib/spack/spack/util/windows_registry.py:
```diff
@@ -366,5 +403,28 @@ def find_value(self, val_name, recursive=True):
+
+class InvalidRegistryOperation(RegistryError):
+    """A Runtime Error ecountered when a registry operation is invalid for
+    a non deterministic reason"""
```
(minor) non-deterministic means "having a different result every time", whereas I think the issue is that when we get these errors we don't have enough info to distinguish what the actual cause is (so it might be the same every time, but we can't tell).

johnwparent replied:
Good catch.

scheibelp commented on lib/spack/spack/operating_systems/windows_os.py:
```diff
@@ -83,11 +85,46 @@ def compiler_search_paths(self):
+
+        def try_query_registry(retry=False):
+            try:
+                # Second strategy: Find MSVC via the registry
```
(minor) I think this specific line would be clearer if moved just above `def try_query_registry`

johnwparent replied:
Good point!

scheibelp commented on lib/spack/spack/operating_systems/windows_os.py:
```diff
@@ -83,11 +85,46 @@ def compiler_search_paths(self):
+                )
+                return msft.find_subkeys(r"VisualStudio_.*", depth=False)
+            except OSError as e:
+                # OSErrors propegated into caller by Spack's registry module are expected
```
(very minor) typo: propagated

scheibelp commented on lib/spack/spack/operating_systems/windows_os.py:
```diff
@@ -83,11 +85,46 @@ def compiler_search_paths(self):
+                    tty.debug(
+                        'Windows registry query on "SOFTWARE\\WOW6432Node\\Microsoft"'
+                        f"under HKEY_LOCAL_MACHINE: {str(e)}"
+                    )
```
(minor) since this is exactly the same as the above error message, I recommend combining into a lambda. Normally I don't mind seeing a message duplicated if it's just twice, but my first thought reading through this is that you were customizing the error message (now I see it's to avoid printing mysterious error messages unless our 2nd retry fails).

scheibelp commented on lib/spack/spack/operating_systems/windows_os.py:
```diff
@@ -83,11 +85,46 @@ def compiler_search_paths(self):
+
+        vs_entries = try_query_registry()
+        if not vs_entries:
+            vs_entries = try_query_registry(retry=True)
```
I think it would be useful to include in a comment here the part of the analysis that suggests retrying immediately is fine (and possibly explaining that retrying once should also be fine).

scheibelp reviewed (changes requested):
I see at least one instance of a docstring referring to a method that was removed in this PR (`_find_subkey_s`) - highlighted in comments.

The rest is mostly suggestions on how to clarify and are generally all minor.

johnwparent committed:
Update comments for more context

johnwparent commented:
@scheibelp thanks for the review! I believe I have addressed all suggestions, ready for another round.

scheibelp commented on lib/spack/spack/operating_systems/windows_os.py:
```diff
@@ -83,11 +85,48 @@ def compiler_search_paths(self):
+                # but the registry raises the same exception for all types of
+                # atypical errors
+                if retry:
+                    winreg_report_error(e)
```
It looks like this can return `None`, can this cause a problem with later `for entry in vs_entries:`?

Sorry for not catching that sooner

johnwparent replied:
No that is intentional, it can return none (i.e. if there's no VS installed), the later check is ~~`for entry in vs_entries` so we're safe if it's none as that will just exit the loop immediately.~~
I'm wrong, it will cause issues, I forgot that python iterables can't take `None` and also that we weren't always returning an empty list.

scheibelp replied:
> No that is intentional, it can return none (i.e. if there's no VS installed)

That part makes sense

> the later check is for entry in vs_entries

What I mean is, if you effectively have:

```
vs_entries = None # sometimes this is the outcome
...
for entry in vs_entries:
    # Iterating over None would fail
```

or am I overlooking something?

johnwparent replied:
Yes, I realized as soon as I hit enter that `None` hadn't really sunk in and I was still thinking in the context of returning an empty list. I've made the fix.

scheibelp commented on lib/spack/spack/util/windows_registry.py:
```diff
@@ -252,11 +299,11 @@ def get_matching_subkeys(self, subkey_name):
             return self.reg.values
 
-    def _traverse_subkeys(self, stop_condition, collect_all_matching=False):
+    def _traverse_subkeys(self, stop_condition, collect_all_matching=False, depth_search=True):
```
Can you update docstring with new arg?

Also, did we agree elsewhere that this concept should generally be represented by the term "recursive" (`find_keys` also introduces `depth`) or do you think this should have a distinct term?

johnwparent replied:
I think we've gone back and forth a few times on this. I think for user facing (non `_` prefixed) methods, we should use recursive as that's the idiom. 
For internal methods, I think it would be a bit off if they had a different arg name, so I'm ok with everything being `recursive`.

scheibelp reviewed (changes requested):
Apologies but I think I missed a couple things in the last review.

johnwparent committed:
review: update winreg docstrings and api

johnwparent committed:
Prevent iterating over None

johnwparent commented:
@scheibelp I believe I've addressed everything

scheibelp commented on lib/spack/spack/util/windows_registry.py:
```diff
@@ -280,67 +328,51 @@ def _traverse_subkeys(self, stop_condition, collect_all_matching=False):
         """
-        kwargs = {"collect_all_matching": True}
-        return self._find_subkey_s(
+        kwargs = {"collect_all_matching": True, "depth_search": recursive}
```
`_traverse_subkeys` now takes `recursive` vs. `depth_search`

Same issue in `find_matching_subkey` and `find_subkey`

johnwparent replied:
🤦

scheibelp reviewed (changes requested):
Last review edits addressed rename of function param but not caller sites.

johnwparent committed:
refactor call sites

johnwparent commented:
@scheibelp I fixed the caller sites in the winreg module and the WindowsOs class, I think that's all of them.

johnwparent committed:
style
